File: easy/Q278_FirstBadVersion.py
# The isBadVersion API is already defined for you.
# @param version, an integer
# @return a bool
def isBadVersion(version):
	return [False, False, False, True, True, True, True][version]


class Solution:
	def firstBadVersion(self, n):
		"""
		:type n: int
		:rtype: int
		"""
		l, mid = 1, (1 + n) // 2
		while l < n:
			if isBadVersion(mid):
				n, mid = mid, (l + mid) // 2
			else:
				l, mid = mid + 1, (n + mid) // 2
		return n


if __name__ == '__main__':
	print(Solution().firstBadVersion(5))

File: src/q278_first_bad_version.rs

pub struct Solution {}

impl Solution {

	// 依赖了外部的 API，不好自己写测试。
	#![allow(non_snake_case)]
	fn isBadVersion(&self, _: i32) -> bool {
		return false;
	}

	pub fn first_bad_version(&self, n: i32) -> i32 {
		let mut s = 0;
		let mut e = n;

		while e > s {
			let mid = s + ((e - s) >> 1);
			if self.isBadVersion(mid) {
				e = mid;
			} else {
				s = mid + 1;
			}
		}
		return s; // 二分题好他妈多啊，果然刷题就是无聊。
	}
}
